### DISCUSSION TITLE
Allow uppercase usernames

### DISCUSSION THREAD

**Discussion by XakepSDK** [keycloak/keycloak#34067] [Ideas]
Origin: https://github.com/keycloak/keycloak/issues/32869

1. Why does the system not allow case-sensitive usernames? Is this because it's troublesome to do case-insensitive search in some databases?
2. Are there plans to support case preservation? I'm thinking about it in this way: username case is preserved, but the name itself must be case-insensitive. If user creates an account with a name `UsEr`, the character case must be preserved. If somebody wants to register an account with a name `user` - it must not be allowed.
3. If you aren't working on it, will you accept contributions? If you are willing to accept it, how you'd like it to be implemented? I'm thinking about adding a second column in the user model that will have case-preserved username and make `getUsername` return it. And lowercase column will be used for all other cases like search, check if user exists or not, etc.

Our workaround is this:
1. We disabled usernames. People are logging in with their email. They don't like it.
2. We use an extension that lets you to force attribute uniqueness
3. We created custom attribute that is the username and we use extensions to enforce uniqueness

This way people can have their username in their favorite case, but they can't use it to log in. We don't like this solution and users hate that they can't log in with their username.

**Comment by codespearhead** [comment]
Can you close this discussion [1] in favor of [2][3]?
 
***

[1] https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions#closing-a-discussion
[2] https://github.com/keycloak/keycloak/discussions/10198
[3] https://github.com/keycloak/keycloak/discussions/10198#discussioncomment-10976212
